feat: introduce `RestrictAccess<To>` and generalize `restrict` to all access types

mkroening opened the pull request:
This PR introduces `trait RestrictAccess<T>` to generalize the restriction from any access type to any other access type, to implement `Volatile{Ptr,Ref}::restrict` generically for all access types.

While not required, this PR also replaces and deprecates `Access::RestrictShared`, since it is equivalent to `RestrictAccess<ReadOnly>::Restricted`.

Though also optional, I implemented `Readable`, `Writable`, and `Copyable` in terms of `RestrictAccess`.

This PR depends on https://github.com/rust-osdev/volatile/pull/59.

mkroening committed:
feat(access): add `RestrictAccess<To>`

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
refactor(ref): implement `borrow` and `as_ptr` via `RestrictAccess`

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
feat(access): deprecate `{Access,Readable}::RestrictShared`

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
feat: generalize `restrict` to all access types

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
refactor(access): implement `Copyable` via `RestrictAccess`

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
refactor(access): implement `Writeable` via `RestrictAccess`

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
refactor(access): implement `Readable` via `RestrictAccess`

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
refactor(access): implement `Access` for `ReadWrite` and `ReadOnly` directly

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
refactor(access): specify `RestrictShared` via ` RestrictAccess`

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

phil-opp reviewed (approved):
Nice abstraction! It makes things a bit more complex, but also more flexible.

I think it would be a good idea to do the breaking release quickly and remove the associated `RestrictShared` type. Otherwise I fear that the two ways to restrict access could lead to confusion. What do you think?

Ready for merge from my side after a rebase.

phil-opp commented on src/access.rs:
```diff
@@ -1,83 +1,108 @@
+///
+/// The following table holds:
+///
+/// | `Self`        | `To`          | `Self` restricted to `To` |
+/// | ------------- | ------------- | ------------------------- |
+/// | `T`           | `T`           | `T`                       |
+/// | [`ReadWrite`] | `T`           | `T`                       |
+/// | [`NoAccess`]  | `T`           | [`NoAccess`]              |
+/// | [`ReadOnly`]  | [`WriteOnly`] | [`NoAccess`]              |
```
I like the table! Maybe we should add rows for `T + NoAccess` and `WriteOnly + ReadOnly` for completeness.

mkroening replied:
Then we'd also need `T` ∩ `ReadWrite`, but this and all the other cases are already covered by the symmetry of the operation. I fear adding these three cases would reduce clarity.

What do you think of the following proposal of highlighting the symmetry more?
I can remove the ∩ symbol again if you'd like, but it might help to highlight the symmetry.

---

# Trait volatile::access::RestrictAccess

A trait for restricting one `Access` type to another `Access` type.

Restricting `Self` to `To` results in `Self::Restricted`.

Restriction is a symmetric operation which is denoted by ∩, as it is the intersection of permissions.
The following table holds:

| `Self`        | `To`          | `Self` ∩ `To` |
| ------------- | ------------- | ------------- |
| `T`           | `T`           | `T`           |
| `ReadWrite`   | `T`           | `T`           |
| `NoAccess`    | `T`           | `NoAccess`    |
| `ReadOnly`    | `WriteOnly`   | `NoAccess`    |

phil-opp replied:
Looks good, thanks!

mkroening committed:
feat(access): remove `RestrictShared`

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening committed:
fix(access): make `Readable`, `Writable`, and `Copyable` bounds consistent

Signed-off-by: Martin Kröning <martin.kroening@eonerc.rwth-aachen.de>

mkroening commented:
> I think it would be a good idea to do the breaking release quickly and remove the associated `RestrictShared` type. Otherwise I fear that the two ways to restrict access could lead to confusion. What do you think?

Works for me. I rebased, removed `RestrictShared` and added a commit for making the bounds on the `Readable`, `Writable`, and `Copyable` traits consistent.

phil-opp reviewed (approved):
Thanks for the updates!
